File: go/backtracking/subsets.go
package main

//子集

func subsets(nums []int) [][]int {
	var res [][]int
	subsetsHelper(nums, 0, []int{}, &res)
	return res
}

func subsetsHelper(nums []int, k int, data []int, res *[][]int) {
	temp := make([]int, len(data))
	copy(temp, data)
	*res = append(*res, temp)

	for i := k; i < len(nums); i++ {
		data = append(data, nums[i])
		subsetsHelper(nums, i+1, data, res)
		data = data[:len(data)-1]
	}
}

File: python/backtracking/subsets.py
from typing import List


# 子集

class Solution:

    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        self.helper(nums, 0, [], res)
        return res
    
    def helper(self, nums, k, data, res):
        res.append(data[:])
        for i in range(k, len(nums)):
            data.append(nums[i])
            self.helper(nums, i + 1, data, res)
            data.pop()
